src/transformStateWithClones.js

```diff
@@ -7,7 +7,27 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    } else if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      const newState = { ...currentState };
 
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+      currentState = newState;
```

Comment:
There's no need to create a new object `newState` here, as you can directly modify `currentState` and then make a copy of it when pushing to `stateHistory`. This will adhere to the DRY principle and reduce unnecessary code.